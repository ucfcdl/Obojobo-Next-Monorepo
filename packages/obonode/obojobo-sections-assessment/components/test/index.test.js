import Dispatcher from 'obojobo-document-engine/src/scripts/common/flux/dispatcher'
import { FOCUS_ON_ASSESSMENT_CONTENT } from '../../assessment-event-constants'
import React from 'react'
import Test from './index'
import renderer from 'react-test-renderer'

jest.mock('obojobo-document-engine/src/scripts/common/flux/dispatcher')
jest.mock('obojobo-document-engine/src/scripts/common/page/focus')

describe('Test', () => {
	beforeEach(() => {
		jest.resetAllMocks()
	})

	test('Test component', () => {
		const model = {
			getComponentClass: jest.fn().mockReturnValueOnce('MockComponent')
		}
		const moduleData = {
			focusState: {}
		}

		const component = renderer.create(<Test model={model} moduleData={moduleData} />)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Test component (isAttemptSubmitting)', () => {
		const model = {
			getComponentClass: jest.fn().mockReturnValueOnce('MockComponent')
		}
		const moduleData = {
			focusState: {}
		}

		const component = renderer.create(
			<Test model={model} moduleData={moduleData} isAttemptSubmitting={true} />
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Test component (isAttemptReadyToSubmit)', () => {
		const model = {
			getComponentClass: jest.fn().mockReturnValueOnce('MockComponent')
		}
		const moduleData = {
			focusState: {}
		}

		const component = renderer.create(
			<Test model={model} moduleData={moduleData} isAttemptReadyToSubmit={true} />
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Test component with completeAttempt', () => {
		const model = {
			getComponentClass: jest.fn().mockReturnValueOnce('MockComponent')
		}
		const moduleData = {
			focusState: {}
		}

		const component = renderer.create(
			<Test model={model} moduleData={moduleData} isAttemptComplete={true} />
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Test component fetching score', () => {
		const model = {
			getComponentClass: jest.fn().mockReturnValueOnce('MockComponent')
		}
		const moduleData = {
			focusState: {}
		}

		const component = renderer.create(
			<Test model={model} moduleData={moduleData} isAttemptComplete={true} isFetching={true} />
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Component listens to FOCUS_ON_ASSESSMENT_CONTENT events when mounted (and stops listening when unmounted)', () => {
		const model = { getComponentClass: jest.fn().mockReturnValueOnce('MockComponent') }
		const moduleData = { focusState: {} }

		expect(Dispatcher.on).not.toHaveBeenCalled()
		expect(Dispatcher.off).not.toHaveBeenCalled()

		const component = renderer.create(<Test model={model} moduleData={moduleData} />)

		const boundFocusOnContent = component.getInstance().boundFocusOnContent
		expect(Dispatcher.on).toHaveBeenCalledTimes(1)
		expect(Dispatcher.on).toHaveBeenCalledWith(FOCUS_ON_ASSESSMENT_CONTENT, boundFocusOnContent)
		expect(Dispatcher.off).not.toHaveBeenCalled()

		component.unmount()

		expect(Dispatcher.on).toHaveBeenCalledTimes(1)
		expect(Dispatcher.off).toHaveBeenCalledTimes(1)
		expect(Dispatcher.off).toHaveBeenCalledWith(FOCUS_ON_ASSESSMENT_CONTENT, boundFocusOnContent)
	})

	test('focusOnContent does nothing (and returns false) when no child model present', () => {
		const model = {
			getComponentClass: jest.fn().mockReturnValueOnce('MockComponent'),
			children: { at: () => null }
		}
		const moduleData = { focusState: {} }

		const component = renderer.create(<Test model={model} moduleData={moduleData} />)

		expect(component.getInstance().focusOnContent()).toBe(false)
	})

	test('focusOnContent does nothing (and returns false) when no child component class exists', () => {
		const model = {
			getComponentClass: jest.fn().mockReturnValueOnce('MockComponent'),
			children: {
				at: () => ({
					getComponentClass: () => false
				})
			}
		}
		const moduleData = { focusState: {} }

		const component = renderer.create(<Test model={model} moduleData={moduleData} />)

		expect(component.getInstance().focusOnContent()).toBe(false)
	})

	test('focusOnContent calls the first child focusOnContent method (and returns true) when child model(s) are present', () => {
		const mockFocusOnContent = jest.fn()
		const mockFirstChild = {
			getComponentClass: () => ({
				focusOnContent: mockFocusOnContent
			})
		}
		const model = {
			getComponentClass: jest.fn().mockReturnValueOnce('MockComponent'),
			children: {
				at: () => mockFirstChild
			}
		}
		const moduleData = { focusState: {} }

		const component = renderer.create(<Test model={model} moduleData={moduleData} />)

		expect(component.getInstance().focusOnContent()).toBe(true)
		expect(mockFocusOnContent).toHaveBeenCalledWith(mockFirstChild)
	})
})
