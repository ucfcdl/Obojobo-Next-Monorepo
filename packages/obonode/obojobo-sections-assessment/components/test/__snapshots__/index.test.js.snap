// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Test component 1`] = `
<div
  className="test"
>
  <MockComponent
    model={
      Object {
        "getComponentClass": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "MockComponent",
            },
          ],
        },
      }
    }
    moduleData={
      Object {
        "focusState": Object {},
      }
    }
  />
  <div
    className="submit-button"
  >
    <div
      className="obojobo-draft--components--button is-not-dangerous align-center"
      contentEditable={false}
    >
      <button
        className="button"
        contentEditable={false}
      >
        Submit (Not all questions have been answered)
      </button>
    </div>
  </div>
</div>
`;

exports[`Test Test component fetching score 1`] = `
<div
  className="test"
>
  <MockComponent
    model={
      Object {
        "getComponentClass": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "MockComponent",
            },
          ],
        },
      }
    }
    moduleData={
      Object {
        "focusState": Object {},
      }
    }
  />
  <div
    className="submit-button"
  >
    <div
      className="obojobo-draft--components--button is-not-dangerous align-center"
      contentEditable={false}
    >
      <button
        className="button"
        contentEditable={false}
        disabled={true}
      >
        Loading ...
      </button>
    </div>
  </div>
</div>
`;

exports[`Test Test component with completeAttempt 1`] = `
<div
  className="test"
>
  <MockComponent
    model={
      Object {
        "getComponentClass": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "MockComponent",
            },
          ],
        },
      }
    }
    moduleData={
      Object {
        "focusState": Object {},
      }
    }
  />
  <div
    className="submit-button"
  >
    <div
      className="obojobo-draft--components--button is-not-dangerous align-center"
      contentEditable={false}
    >
      <button
        className="button"
        contentEditable={false}
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;
